### PR TITLE
Fix ComputeUncompute State Fidelity threading issue

### LINKED ISSUE #91: ComputeUncompute State Fidelity threading issue
### Description

Recently [this job failed](https://github.com/qiskit-community/qiskit-algorithms/actions/runs/6500429199/job/17655786517?pr=88), as below, in  similar manner to that raised by
* Qiskit/qiskit#10179

Looking into this further its not the test case as such rather the ComputeUncompute implementation and the job thread logic it uses, more detail below.

```
test.state_fidelities.test_compute_uncompute.TestComputeUncompute.test_symmetry
-------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/qiskit_algorithms/state_fidelities/compute_uncompute.py", line 161, in _run
    result = job.result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/qiskit/primitives/primitive_job.py", line 55, in result
    return self._future.result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/qiskit/primitives/sampler.py", line 88, in _call
    if len(value) != len(self._parameters[i]):

    IndexError: list index out of range

    
The above exception was the direct cause of the following exception:


    Traceback (most recent call last):

      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/test/state_fidelities/test_compute_uncompute.py", line 106, in test_symmetry
    results_2 = job_2.result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/qiskit/primitives/primitive_job.py", line 55, in result
    return self._future.result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception

      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)

      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/qiskit_algorithms/state_fidelities/compute_uncompute.py", line 163, in _run
    raise AlgorithmError("Sampler job failed!") from exc

    qiskit_algorithms.exceptions.AlgorithmError: 'Sampler job failed!'
```    

### Reproducing the error

Running this script, which replicates the logic the failing test, fairly quickly fails the same way as show above.

```
import numpy as np

from qiskit.circuit import QuantumCircuit, ParameterVector
from qiskit.primitives import Sampler

from qiskit_algorithms.state_fidelities import ComputeUncompute


def main():
    i = 1
    while True:
        do_fidelity()
        print(f"{i}\r", end="")
        i+=1


def do_fidelity(serial=False):

    parameters = ParameterVector("x", 2)

    circuit = QuantumCircuit(2)
    circuit.rx(parameters[0], 0)
    circuit.rx(parameters[1], 1)

    sampler = Sampler()
    params_l = np.array([[0, 0], [np.pi / 2, 0], [0, np.pi / 2], [np.pi, np.pi]])
    params_r = np.array([[0, 0], [0, 0], [np.pi / 2, 0], [0, 0]])

    fidelity = ComputeUncompute(sampler)
    n = len(params_l)
    job_1 = fidelity.run(
        [circuit] * n, [circuit] * n, params_l, params_r
    )
    job_2 = fidelity.run(
        [circuit] * n, [circuit] * n, params_r, params_l
    )
    results_1 = job_1.result()
    results_2 = job_2.result()

    if not np.allclose(results_1.fidelities, results_2.fidelities, atol=1e-16):
        print("Failed")
        sys.exit()


if __name__ == "__main__":
    main()
```

### Analysis

From investigation it appears that failure is that the `sampler.run()` is part of the logic called by the function that is passed job thread producing the result. That sampler code updates a number of instance variables and is not safe for such usage. I verified things by acquiring and releasing a threading lock in the fidelity primitive around the run call and the above has run over a weekend with no failure. That will not do as a general solution though.

### Suggestion

My suggestion would be, if we want the same async behavior on the result/job, to change what is done in the thread to just be the call to the get the sampler job result, from running the sampler (which should happen in the main thread instead) and format up the result. It will be far less in the thread but that should be safe.

### PR DESCRIPTION
**Pull request by woodsp-ibm**
### Summary

Fixes #91

- [x] Add reno (as noted below)

### Details and comments

I changed the logic so that the `sampler.run` is done in the main thread and the resulting job and result is processed for the fidelity result in a separate method that is run by the thread executor of the algorithm job. This new method (_call which I named to be the same as the ref. primitives) gets passed all it needs as params, self is not passed, and where that logic was using some private instance utility methods I turned these to static (they never used self anyway).

As far as an end user goes the ComputeUncompute is used just the same way. As I changed things around a little bit though any developer that might have extended the base class, for some custom fidelity, would need to update their code for the _run. 

I kept the changes minimal though the presence of the _run is a little harder to justify as things are now, but I figure, if more fidelities are added going forwards that may be a better time to sort out commonality than change things around more now.

I'll add a reno once others have had a chance to look over this. It would note the bug fix and point out the change to _run should anyone happen to have made their own fidelity off the base.

### CODE REVIEW

File: qiskit_algorithms/state_fidelities/compute_uncompute.py [hunk 1]
```diff
@@ -135,7 +137,7 @@ def _run(
                     Higher priority setting overrides lower priority setting.
 
         Returns:
-            The result of the fidelity calculation.
+            An AlgorithmJOb for the fidelity calculation.
```
Comment (ElePT) [inline qiskit_algorithms/state_fidelities/compute_uncompute.py:140]:
```suggestion
            An AlgorithmJob for the fidelity calculation.
```

**Review by ElePT — COMMENTED** [review]
Thank you very much for looking into this. I now know that the use of `self` is not a very good friend of multi-threading, but didn't see it as an issue at the time. As also mentioned in #93, I think that the primitives interface will change anyways, so I am ok with keeping the current `_run` as-is and revisiting the fidelities architecture once the primitives are settled. Unless @Cryoris has any issue with it, I would go ahead with the reno.

**Comment by woodsp-ibm** [comment]
Its not so much about self itself more that it refers to a mutable object and if the thread can read values that are changing in the main thread context... Passing whatever values that the function run under the thread needs to me makes things explicit rather than passing a reference to a larger object and letting it have at it, and the values are on the stack. To me its more playing it safe and hopefully, if any future changes might be made, doing things explicitly like this is clearer etc. Of course if one of those values, is really a reference and again mutable you are back towards passing self....

**Review by ElePT — APPROVED** [review]
I think that we can go ahead and merge the fix to reduce the flakiness of the tests. And maybe reconsider if any objections come up in the future.

**Pull request merged**